File: dynamic-programming/grid/minimum-path-sum/java/Solution_StateCompressionDp.java
import java.util.Arrays;

class Solution_StateCompressionDp {
    public int minPathSum(int[][] grid) {
        int n = grid[0].length;
        int[] current = new int[n + 1];
        Arrays.fill(current, Integer.MAX_VALUE);
        current[1] = 0;
        for (int[] row : grid) {
            for (int i = 1; i <= n; i++) {
                current[i] = Math.min(current[i - 1], current[i]) + row[i - 1];
            }
        }
        return current[n];
    }
}

File: dynamic-programming/grid/minimum-path-sum/ts/solution_state-compression-dp.ts
function minPathSum(grid: number[][]): number {
    const n = grid[0].length
    const current: number[] = Array(n + 1).fill(Number.MAX_SAFE_INTEGER)
    current[1] = 0
    for (let row of grid) {
        for (let i = 1; i <= n; i++) {
            current[i] = Math.min(current[i - 1], current[i]) + row[i - 1]
        }
    }
    return current[n]
}

export { minPathSum }